### PR TITLE
`struct Rav1dFrameContext_frame_thread`: Add inner mutability

### PR DESCRIPTION
**Pull request by rinon**

### CODE REVIEW

**Review by randomPoison — APPROVED** [review]
Changes mostly look good, I just had a couple of minor comments.

File: src/recon.rs
Comment (randomPoison) [inline src/recon.rs:1627]: Now that I think about it, it's probably not a good idea to pass a `&mut Rav1dTaskContext` here just to use one field, i.e. you might want to take `t.cf` as the argument, rather than all of `t`. The current usages of this are dereferencing a `*mut Rav1dTaskContext` so you're not hitting borrow errors yet, but once `decode_coefs` is taking a `t: &mut Rav1dTaskContext` you're likely going to hit borrow errors trying to take a mutable reference to all of `t`.

Comment (randomPoison) [reply src/recon.rs:1627]: The same also probably applies to `get`, I think.

Comment (rinon) [reply src/recon.rs:1627]: Probably, but it's easy enough to fix IF we hit that, rather than making this a bit less clear preemptively IMO.

File: src/recon.rs
Comment (rinon) [reply src/recon.rs:1624]: Making `set` unsafe, I think get can be safe (see discussion on #890 )

File: src/decode.rs
Comment (kkysen) [inline src/decode.rs:1161]:
```suggestion
```

File: src/recon.rs
Comment (kkysen) [reply src/recon.rs:1627]: I agree it'd be better to take `cf: {&,&mut} BitDepthUnion<Cf>` as an individual field here instead of all of `t`.  And I think it's clearer this way as well.

File: src/internal.rs
Comment (kkysen) [inline src/internal.rs:514]: This increases memory usage by 25% (32 vs. 40 bytes).  I think it's better to make the fields atomic, as they're generally accessed individually from what I can remember.  And we also get rid of the `Mutex` overhead.  I can open an issue for this.

**Review by kkysen — CHANGES_REQUESTED** [review]

File: src/internal.rs
Comment (rinon) [reply src/internal.rs:551]: I think it makes more sense to access the block as a whole rather than add atomicity inside. We don't access blocks concurrently afaict. I'll make this vec DisjointMut for now to avoid the memory bloat.

Comment (kkysen) [reply src/internal.rs:551]: I was about to work on making `Av1Block` use `enum`s, which is a noisy but not super complex change.  Should I hold off on that to avoid conflicts with this then?

File: src/recon.rs
Comment (rinon) [reply src/recon.rs]: Didn't see any easy way, but now that we support slices over buffers properly we can. Fixed.

File: src/internal.rs
Comment (rinon) [reply src/internal.rs]: Yes please, don't do that, that's not necessary.

Comment (kkysen) [reply src/internal.rs]: Okay I can wait, but what do you mean it's not necessary?  We need to remove the `union`s.

Comment (rinon) [reply src/internal.rs]: I'm sorry, I read that as atomics. I should read more carefully when I'm tired. Just make sure you're coordinating with @randomPoison to ensure you're not both planning to do that. I think it is best to wait until this lands though, since you haven't done it yet.

File: include/common/bitdepth.rs
Comment (kkysen) [inline include/common/bitdepth.rs:244]: Do we still need this?  It seems unused now.

**Review by kkysen — CHANGES_REQUESTED** [review]
Overall LGTM, though a few minor things:
* I think `cast_coef_mut` isn't used now.  Do we intend on using it now that we have `.slice_as`, etc.?
* One place we added extra `as isize` casts that aren't needed.
* I'd prefer to take `cf: {&,&mut} BitDepthUnion<Cf>` directly instead all of `t` in `get` and `set.
* A bunch of places where we slice and then add the length to the offset, you added the new length as `len`, which duplicates the existing inline `+=` to the offset.  If we could use `len` in both places, that'd be perfect.

File: src/recon.rs
Comment (rinon) [reply src/recon.rs:3729]: If that is the case then it probably shouldn't be `ptrdiff_t`. I'm leaving this to match other usage unless and until we change that.

Comment (kkysen) [reply src/recon.rs:3724]: The other usages don't do the `isize` cast, though, only this one.

**Pull request merged**
